### PR TITLE
search autocomplete in js

### LINKED ISSUE #44: Provide a list of recent local searches for the user to select from when interacting with the search box
We would like users to be able to select, from a dropdown below the top navigation search box, recent searches.

image

image

* The search terms should be stored in local storage and therefore no backend support is required.
* Initially, when there is no search term entered, the last n searches should be displayed
* Once there is a search term (or part of a term), perform an "includes" filter on the list and only display the most recent matching n searches. Extra points if we can highlight the match within the search.
* This should work _without_ the ClojureScript app and therefore the solution to be written in JavaScript (there are plenty of examples of this within the project)
* Extra points for allowing the searches to be deleted

### PR DESCRIPTION
**Pull request by fctorial**
Closes #44 . 

Note: I had to install 'react-modal' in order to start working on the master branch. 'npm run dev'  would fail without it.

### CODE REVIEW

**Comment by fctorial** [comment]
The new file `ac.js` contains some variables that can be used for tweaking.

File: client/src-js/ac.js [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+
+function hideMenu(n) {
+  const e = n.parentNode.querySelector('.'+WRAPPER_CLASS)
+  e && e.remove()
+}
+
+function onSearchKey(e) {
+  const ip = e.target
+  let history = getHistory()
+  console.log(e.key)
```
Comment (jonki) [inline client/src-js/ac.js:40]: Probably not required anymore ;)

File: client/src-js/ac.js [hunk 2]
```diff
@@ -0,0 +1,159 @@
+const HISTORY_SIZE = 20
+const HISTORY_LIST_SIZE = 4
+const DB_ID = "search_history"
+const WRAPPER_CLASS = "search_class"
+const HOVER_COLOR = '#f4ada4'
+const HOVER_COLOR_CANCEL = '#dddddd'
+const BG_COLOR = '#fefefe'
+const PRIMARY = '#ef8374'
+
```
Comment (jonki) [inline client/src-js/ac.js:9]: appreciate creating this config

File: client/src-js/ac.js [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+  let res = []
+  for (let i=0; i<n; i++) {
+    res.push(i)
+  }
+  return res
+}
+
+function getHistory() {
+  try {
+    return JSON.parse(localStorage.getItem(DB_ID))
```
Comment (jonki) [inline client/src-js/ac.js:20]: You need something like `return JSON.parse(localStorage.getItem(DB_ID)) || []` to guard this function from returning `null` before any history is saved in storage

File: client/src-js/ac.js [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+  localStorage.setItem(DB_ID, JSON.stringify(history))
+}
+
+function hideMenu(n) {
+  const e = n.parentNode.querySelector('.'+WRAPPER_CLASS)
+  e && e.remove()
+}
+
+function onSearchKey(e) {
+  const ip = e.target
```
Comment (jonki) [inline client/src-js/ac.js:38]: Not sure why you call event target `ip`? Is it supposed to be `input`? If that's the case I would simply use `input`. Also relevant in other event handlers

**Review by jonki — COMMENTED** [review]
Looks good. I've added few little comments, I'd appreciate if you could address them. I need to discuss your solution with my colleagues, but at first glance seems pretty solid and fulfilling our requirements. Thanks so much!

**Comment by yanisurbis** [comment]
Hello, really impressive work!

One last thing before me merge this PR :)

I see that a lot of css values are copied. May we use classes instead? It's true that css modules prefix all classes.
There is a way to create unprefixed classes though.

image

This code will generate 3 classes which work inside our navbar
image

So you are able to reuse them in your code without copying all css values in your code
image

File: client/src-js/ac.js [hunk 5]
```diff
@@ -0,0 +1,159 @@
+const HISTORY_SIZE = 20
+const HISTORY_LIST_SIZE = 4
+const DB_ID = "search_history"
+const WRAPPER_CLASS = "search_class"
+const HOVER_COLOR = '#f4ada4'
```
Comment (yanisurbis) [inline client/src-js/ac.js:5]: 1

**Review by yanisurbis — CHANGES_REQUESTED** [review]

**Review by acron0 — APPROVED** [review]

**Comment by acron0** [comment]
OK TO MERGE

**Comment by acron0** [comment]
@fctorial Can you send me an email antony@functionalworks.com please, thanks :)

**Comment by acron0** [comment]
OK TO MERGE

**Comment by acron0** [comment]
OK TO MERGE

**Comment by acron0** [comment]
This was merged upstream in https://github.com/WorksHub/works-hub/pull/3037

**Pull request closed without merging**